balancer/pickfirstleaf/test/pickfirstleaf_test.go

```diff
@@ -0,0 +1,731 @@
+			name:                "two_server_first_ready",
+			backendIndexes:      []int{0, 1},
+			runningBackendIndex: 0,
+			wantSCStates: []scStateExpectation{
+				{ServerIdx: 0, State: connectivity.Ready},
+			},
+			wantConnStateTransitions: []connectivity.State{
+				connectivity.Connecting,
+				connectivity.Ready,
+			},
+		},
+		{
+			name:                "two_server_second_ready",
+			backendIndexes:      []int{0, 1},
+			runningBackendIndex: 1,
+			wantSCStates: []scStateExpectation{
+				{ServerIdx: 0, State: connectivity.Shutdown},
+				{ServerIdx: 1, State: connectivity.Ready},
+			},
+			wantConnStateTransitions: []connectivity.State{
+				connectivity.Connecting,
+				connectivity.Ready,
+			},
+		},
+		{
+			name:                "duplicate_address",
+			backendIndexes:      []int{0, 0, 1},
+			runningBackendIndex: 1,
+			wantSCStates: []scStateExpectation{
+				{ServerIdx: 0, State: connectivity.Shutdown},
+				{ServerIdx: 1, State: connectivity.Ready},
+			},
+			wantConnStateTransitions: []connectivity.State{
+				connectivity.Connecting,
+				connectivity.Ready,
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			cc, r, bm := setupPickFirstLeaf(t, 2)
+			addrs := stubBackendsToResolverAddrs(bm.backends)
+			stateSubscriber := &ccStateSubscriber{transitions: []connectivity.State{}}
+			internal.SubscribeToConnectivityStateChanges.(func(cc *grpc.ClientConn, s grpcsync.Subscriber) func())(cc, stateSubscriber)
+
+			activeAddrs := []resolver.Address{}
+			for _, idx := range tc.backendIndexes {
+				activeAddrs = append(activeAddrs, addrs[idx])
+			}
+
+			bm.stopAllExcept(tc.runningBackendIndex)
+			runningAddr := addrs[tc.runningBackendIndex]
+
+			r.UpdateState(resolver.State{Addresses: activeAddrs})
+			bal := <-balChan
+			testutils.AwaitState(ctx, t, cc, connectivity.Ready)
+
+			if err := pickfirst.CheckRPCsToBackend(ctx, cc, runningAddr); err != nil {
+				t.Fatal(err)
+			}
+
+			if diff := cmp.Diff(scStateExpectationToSCState(tc.wantSCStates, addrs), bal.subConns()); diff != "" {
+				t.Errorf("subconn states mismatch (-want +got):\n%s", diff)
+			}
+
+			if diff := cmp.Diff(tc.wantConnStateTransitions, stateSubscriber.transitions); diff != "" {
+				t.Errorf("ClientConn states mismatch (-want +got):\n%s", diff)
+			}
+		})
+	}
+}
+
+// TestPickFirstLeaf_ResolverUpdate tests the behaviour of the pick first
+// policy when the following steps are carried out in order:
+//  1. A list of addresses are given through the resolver. Only one
+//     of the servers is running.
+//  2. RPCs are sent to verify they reach the running server.
+//  3. A second resolver update is sent. Again, only one of the servers is
+//     running. This may not be the same server as before.
+//  4. RPCs are sent to verify they reach the running server.
+//
+// The state transitions of the ClientConn and all the subconns created are
+// verified.
+func (s) TestPickFirstLeaf_MultipleResolverUpdates(t *testing.T) {
```

Comment:
Similar comments for this test too. If we can get each individual test to be less than 50 lines of code (arbitrary) such that we can read through the whole thing without having to scroll up and down, i would still go ahead and break this table into individual tests.

Reply:
Removed table driven tests.